<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Read alignment - rRNA maturation</title><link rel="shortcut icon" href="/rRNA_maturation/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/rRNA_maturation/assets/css/just-the-docs-default.css"> <script src="/rRNA_maturation/assets/js/vendor/lunr.min.js"></script> <script src="/rRNA_maturation/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Read alignment | rRNA maturation</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Read alignment" /><meta property="og:locale" content="en_US" /><meta name="description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><meta property="og:description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><link rel="canonical" href="http://localhost:4000/rRNA_maturation/processing_analysis/read_alignment/" /><meta property="og:url" content="http://localhost:4000/rRNA_maturation/processing_analysis/read_alignment/" /><meta property="og:site_name" content="rRNA maturation" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Read alignment" /> <script type="application/ld+json"> {"@type":"WebPage","url":"http://localhost:4000/rRNA_maturation/processing_analysis/read_alignment/","headline":"Read alignment","description":"Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/rRNA_maturation/" class="site-title lh-tight"> rRNA maturation </a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="/rRNA_maturation/" class="nav-list-link">About</a><li class="nav-list-item"><a href="/rRNA_maturation/project" class="nav-list-link">Project</a><li class="nav-list-item"><a href="/rRNA_maturation/sample_preparation/" class="nav-list-link">Sample preparation</a><li class="nav-list-item"><a href="/rRNA_maturation/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item active"><a href="#" class="nav-list-expander" aria-label="toggle links in Processing analysis category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/processing_analysis/" class="nav-list-link">Processing analysis</a><ul class="nav-list "><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/Basecalling/" class="nav-list-link">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item active"><a href="/rRNA_maturation/processing_analysis/read_alignment/" class="nav-list-link active">Read alignment</a></ul><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Modified base detection category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/mod_base/" class="nav-list-link">Modified base detection</a><ul class="nav-list "></ul><li class="nav-list-item"><a href="/rRNA_maturation/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="/rRNA_maturation/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search rRNA maturation" aria-label="Search rRNA maturation" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/rRNA_maturation" class="site-button" > Workflow on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><nav aria-label="Breadcrumb" class="breadcrumb-nav"><ol class="breadcrumb-nav-list"><li class="breadcrumb-nav-list-item"><a href="http://localhost:4000/rRNA_maturation/processing_analysis/">Processing analysis</a><li class="breadcrumb-nav-list-item"><span>Read alignment</span></ol></nav><div id="main-content" class="main-content" role="main"><h2 class="no_toc" id="read-alignment"> <a href="#read-alignment" class="anchor-heading" aria-labelledby="read-alignment"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Read alignment</h2><hr /> <details open=""> <summary class="text-delta"> Table of contents </summary><ol id="markdown-toc"><li><a href="#genome-files" id="markdown-toc-genome-files">Genome files</a><li><a href="#mapping-of-reads-to-the-genome-using-minimap2" id="markdown-toc-mapping-of-reads-to-the-genome-using-minimap2">Mapping of reads to the genome using <code class="language-plaintext highlighter-rouge">minimap2</code></a></ol></details><hr /><h3 id="genome-files"> <a href="#genome-files" class="anchor-heading" aria-labelledby="genome-files"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Genome files</h3><ul><li>Genome FASTA and GFF3 files have been downloaded from <a href="https://www.ncbi.nlm.nih.gov/nuccore/545778205">GenBank</a><li>Transcript file was made using <a href="https://github.com/gpertea/gffread"><code class="language-plaintext highlighter-rouge">gffread</code></a> with <code class="language-plaintext highlighter-rouge">gffread microbepore/data/genome/NC_000913.3.gff -g microbepore/data/genome/NC_000913.3.fasta -w microbepore/data/genome/NC_000913.3.transcripts.fasta</code></ul><h3 id="mapping-of-reads-to-the-genome-using-minimap2"> <a href="#mapping-of-reads-to-the-genome-using-minimap2" class="anchor-heading" aria-labelledby="mapping-of-reads-to-the-genome-using-minimap2"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Mapping of reads to the genome using <a href="https://github.com/lh3/minimap2"><code class="language-plaintext highlighter-rouge">minimap2</code></a></h3><ul><li>Files were mapped to the reference genome from <em>Escherichia coli</em> K-12 MG1655 (<a href="https://www.ncbi.nlm.nih.gov/nuccore/545778205">GenBank</a>: U00096.3) using <code class="language-plaintext highlighter-rouge">minimap2</code> (Release 2.18-r1015)<li>Output alignments in the SAM format were generated with <code class="language-plaintext highlighter-rouge">-ax splice -k14</code> for <strong>Nanopore cDNA-seq</strong> and <code class="language-plaintext highlighter-rouge">-ax splice, -uf, -k14</code> for <strong>DRS</strong> with i) <code class="language-plaintext highlighter-rouge">-p 0.99</code>, to return primary and secondary mappings and ii) with <code class="language-plaintext highlighter-rouge">--MD</code>, to include the MD tag for calculating mapping identities<li>Alignment files were further converted to BAM files, sorted and indexed using [<code class="language-plaintext highlighter-rouge">SAMtools</code>(https://github.com/samtools/)<li>To analyse single reads in more detail with respect to the RNA type (mRNA, rRNA, other ncRNA, unspecified) they map to, BAM files were first converted back to FASTQ using <a href="https://bedtools.readthedocs.io/en/latest/"><code class="language-plaintext highlighter-rouge">bedtools</code></a> v2.29.2<li>Next FASTQ files were remapped to a transcriptome file using <code class="language-plaintext highlighter-rouge">minimap2</code> with the previously mentioned parameters to assign single read names with feature IDs</ul><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/FASTQ/normal <span class="c"># input directory with all merged FASTQ files, 1 for each barcode or single DRS run</span>
<span class="nv">fasta</span><span class="o">=</span>microbepore/data/genome/NC_000913.3.fasta <span class="c"># downloaded from GenBank</span>
<span class="nv">transcripts</span><span class="o">=</span>microbepore/data/genomeNC_000913.3.transcripts.fasta <span class="c"># transcripts file made using gffread</span>

<span class="c"># Mapping &amp; Remapping - loop through all FASTQs</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>.fastq
<span class="k">do</span>
  
  <span class="c"># folder and filenames</span>
  <span class="nv">f_ex</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="k">${</span><span class="nv">f_ex</span><span class="k">}</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span> <span class="c"># depending on how you name your files </span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">f_ex</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="c"># make directories</span>
  <span class="nb">mkdir </span>microbepore/data/mapped/raw <span class="c"># direct output to mapped folder for raw reads</span>
  <span class="nb">mkdir </span>microbepore/data/mapped/raw/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span> <span class="c"># run_id</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/mapped/raw/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span> <span class="c"># run_id/barcode_id</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$filename</span> <span class="o">=</span>~ <span class="s2">"RNA"</span> <span class="o">]]</span><span class="p">;</span> 
  <span class="k">then</span>
  <span class="c"># align using minimap2</span>
  minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-uf</span> <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">fasta</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sam <span class="c"># DRS</span>
  <span class="k">else
    </span>minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">fasta</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sam <span class="c"># (PCR-)cDNA</span>
  <span class="k">fi</span>
 
  <span class="c"># convert to sorted.bam file</span>
  samtools view <span class="nt">-bS</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.bam
  samtools <span class="nb">sort</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.bam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sorted.bam
  samtools index <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sorted.bam
  
  <span class="c"># bam to fastq for remapping of mapped reads</span>
  bedtools bamtofastq <span class="nt">-i</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sorted.bam <span class="nt">-fq</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.fastq
  
  <span class="c"># map again</span>
  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$filename</span> <span class="o">=</span>~ <span class="s2">"RNA"</span> <span class="o">]]</span><span class="p">;</span> 
  <span class="k">then
  </span>minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-uf</span> <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">transcripts</span><span class="k">}</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.fastq <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam
  <span class="k">else
    </span>minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">transcripts</span><span class="k">}</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.fastq <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam
 <span class="k">fi</span>
 
  <span class="c"># convert to sorted.bam file</span>
  samtools view <span class="nt">-bS</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.bam
  samtools <span class="nb">sort</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.bam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sorted.bam
  samtools index <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sorted.bam
<span class="k">done</span>
</code></pre></div></div><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2022 Felix Grünberger. Distributed by an <a href="https://github.com/felixgrunberger/rRNA_maturation/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
