<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Detection of transcript boundaries - microbepore</title><link rel="shortcut icon" href="/microbepore/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/microbepore/assets/css/just-the-docs-default.css"> <script type="text/javascript" src="/microbepore/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="/microbepore/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Detection of transcript boundaries | microbepore</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Detection of transcript boundaries" /><meta property="og:locale" content="en_US" /><meta name="description" content="Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules" /><meta property="og:description" content="Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules" /><link rel="canonical" href="http://localhost:4000/microbepore/data_analysis/transcript_boundaries/" /><meta property="og:url" content="http://localhost:4000/microbepore/data_analysis/transcript_boundaries/" /><meta property="og:site_name" content="microbepore" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Detection of transcript boundaries" /> <script type="application/ld+json"> {"description":"Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules","headline":"Detection of transcript boundaries","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"http://localhost:4000/microbepore/assets/images/microbepore_logo.png"}},"url":"http://localhost:4000/microbepore/data_analysis/transcript_boundaries/","@type":"WebPage","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/microbepore/" class="site-title lh-tight"><div class="site-logo"></div></a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="http://localhost:4000/microbepore/" class="nav-list-link">About</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/project" class="nav-list-link">Project</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/sample_preparation/" class="nav-list-link">Sample preparation</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item active"><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="http://localhost:4000/microbepore/data_analysis/" class="nav-list-link">Data analysis</a><ul class="nav-list "><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/project_organization/" class="nav-list-link">Project organisation</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/Basecalling/" class="nav-list-link">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/read_alignment/" class="nav-list-link">Read alignment</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/identification_full_length/" class="nav-list-link">Identification and trimming of full-length reads</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/quantification/" class="nav-list-link">Gene abundance estimation</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/qc_raw_reads/" class="nav-list-link">Quality control</a><li class="nav-list-item active"><a href="http://localhost:4000/microbepore/data_analysis/transcript_boundaries/" class="nav-list-link active">Detection of transcript boundaries</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/gene_body/" class="nav-list-link">Gene body coverage</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/operon_detection/" class="nav-list-link">Transcriptional unit analysis</a></ul><li class="nav-list-item"><a href="http://localhost:4000/microbepore/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search microbepore" aria-label="Search microbepore" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/microbepore" class="site-button" > microbepore on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><nav aria-label="Breadcrumb" class="breadcrumb-nav"><ol class="breadcrumb-nav-list"><li class="breadcrumb-nav-list-item"><a href="http://localhost:4000/microbepore/data_analysis/">Data analysis</a><li class="breadcrumb-nav-list-item"><span>Detection of transcript boundaries</span></ol></nav><div id="main-content" class="main-content" role="main"><h2 class="no_toc" id="detection-of-transcript-boundaries"> <a href="#detection-of-transcript-boundaries" class="anchor-heading" aria-labelledby="detection-of-transcript-boundaries"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Detection of transcript boundaries</h2><hr /> <details open=""> <summary class="text-delta"> Table of contents </summary><ol id="markdown-toc"><li><a href="#5end-detection" id="markdown-toc-5end-detection">5´end detection</a><li><a href="#3end-detection" id="markdown-toc-3end-detection">3´end detection</a></ol></details><hr /><p>The determination of enriched 5´and 3´ends was carried out in the same way, but independently of each other, and is briefly explained in the following:</p><ul><li>First, strand-specific read ends in bedgraph format were created from BAM files using <a href="https://bedtools.readthedocs.io/en/latest/"><code class="language-plaintext highlighter-rouge">bedtools genomecov</code></a> (-5 or -3 option, -bga)<li>Next, the previously published <a href="https://pypi.org/project/termseq-peaks/"><code class="language-plaintext highlighter-rouge">Termseq_peaks</code></a> script was used to call peaks for each sample individually without including replicates (https://github.com/NICHD-BSPC/termseq-peaks)<li>This script is based on <code class="language-plaintext highlighter-rouge">scipy.signal.find_peaks</code>, which is running in the background of <code class="language-plaintext highlighter-rouge">Termseq_peaks</code> with lenient parameters (prominence=(None,None), width=(1,None), rel_height=0.75)<li>However, we deliberately used <code class="language-plaintext highlighter-rouge">Termseq_peaks</code> since its ability to include replicates by applying an Irreproducible Discovery Rate method which can be applied to future studies<li>For end detection, only the leniently called peaks in the narrowPeak file were used after adding the number of counts for each position using <code class="language-plaintext highlighter-rouge">bedtools intersect</code>.</ul><h3 id="5end-detection"> <a href="#5end-detection" class="anchor-heading" aria-labelledby="5end-detection"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> 5´end detection</h3><p>5´end peak calling was performed in the following way:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/mapped

<span class="c"># perform tss detection for pychopper auto &gt; cutadapt_polyA &gt; SSP-cutadapt &gt; clipped  or for raw mapped reads</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/trimmed/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>clipped.sorted.bam <span class="c"># ||  for file in ${input}/raw/*/*/*.sorted.bam</span>
<span class="k">do</span> 
  <span class="c"># file and folder names</span>
  <span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"."</span> <span class="nt">-f</span> 2<span class="si">)</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="c"># make directories</span>
  <span class="nb">mkdir </span>microbepore/data/tss/trimmed
  <span class="nb">mkdir </span>microbepore/data/tss/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/tss/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  <span class="c"># step 1: calculate 5´positions for plus and minus strand</span>
  bedtools genomecov <span class="se">\</span>
    <span class="nt">-ibam</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">-bga</span> <span class="se">\</span>
    <span class="nt">-5</span> <span class="se">\</span>
    <span class="nt">-strand</span> + <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph
  
  bedtools genomecov <span class="se">\</span>
    <span class="nt">-ibam</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">-bga</span> <span class="se">\</span>
    <span class="nt">-5</span> <span class="se">\</span>
    <span class="nt">-strand</span> - <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph
    
  <span class="c"># step 2: termseq peaks</span>
  termseq_peaks <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="nt">--peaks</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks <span class="nt">--strand</span> +
  termseq_peaks <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="nt">--peaks</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks <span class="nt">--strand</span> -
    
  <span class="c"># step 3: add coverage information</span>
  bedtools intersect <span class="se">\</span>
    <span class="nt">-wao</span> <span class="se">\</span>
    <span class="nt">-a</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks.oracle.narrowPeak <span class="se">\</span>
    <span class="nt">-b</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="se">\</span>
    <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks.oracle.narrowPeak.counts
    
  bedtools intersect <span class="se">\</span>
    <span class="nt">-wao</span> <span class="se">\</span>
    <span class="nt">-a</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks.oracle.narrowPeak <span class="se">\</span>
    <span class="nt">-b</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="se">\</span>
    <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks.oracle.narrowPeak.counts

<span class="k">done</span>
</code></pre></div></div><h3 id="3end-detection"> <a href="#3end-detection" class="anchor-heading" aria-labelledby="3end-detection"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> 3´end detection</h3><p>3´end peak calling was performed in the following way:</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/mapped

<span class="c"># perform tts detection for pychopper auto &gt; cutadapt_polyA &gt; SSP-cutadapt &gt; clipped  or for raw mapped reads</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/trimmed/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>clipped.sorted.bam <span class="c"># ||  for file in ${input}/raw/*/*/*.sorted.bam</span>
<span class="k">do 
  </span><span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"."</span> <span class="nt">-f</span> 2<span class="si">)</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
    
  <span class="nb">echo</span> <span class="k">${</span><span class="nv">filename</span><span class="k">}</span>

  <span class="nb">mkdir </span>microbepore/data/tts/trimmed
  <span class="nb">mkdir </span>microbepore/data/tts/trimmed
  <span class="nb">mkdir </span>microbepore/data/tts/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/tts/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  <span class="c"># step 1: calculate 3´positions for plus and minus strand</span>
  bedtools genomecov <span class="se">\</span>
    <span class="nt">-ibam</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">-bga</span> <span class="se">\</span>
    <span class="nt">-3</span> <span class="se">\</span>
    <span class="nt">-strand</span> + <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph
  
  bedtools genomecov <span class="se">\</span>
    <span class="nt">-ibam</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="se">\</span>
    <span class="nt">-bga</span> <span class="se">\</span>
    <span class="nt">-3</span> <span class="se">\</span>
    <span class="nt">-strand</span> - <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph
    
  <span class="c"># step 2: termseq peaks</span>
  termseq_peaks <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="nt">--peaks</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks <span class="nt">--strand</span> +
    termseq_peaks <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="nt">--peaks</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks <span class="nt">--strand</span> -
    
  <span class="c"># step 3: add coverage information</span>
  bedtools intersect <span class="se">\</span>
    <span class="nt">-wao</span> <span class="se">\</span>
    <span class="nt">-a</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks.oracle.narrowPeak <span class="se">\</span>
    <span class="nt">-b</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.bedgraph <span class="se">\</span>
    <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.plus.peaks.oracle.narrowPeak.counts
    
  bedtools intersect <span class="se">\</span>
    <span class="nt">-wao</span> <span class="se">\</span>
    <span class="nt">-a</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks.oracle.narrowPeak <span class="se">\</span>
    <span class="nt">-b</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.bedgraph <span class="se">\</span>
    <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.minus.peaks.oracle.narrowPeak.counts

<span class="k">done</span>
</code></pre></div></div><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2021 Felix Grünberger. Distributed by an <a href="https://github.com/felixgrunberger/microbepore/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
