Insights into rRNA processing and modifications in Archaea using
Nanopore-based RNA sequencing
================
<a href="https://orcid.org/0000-0001-7444-2408">Felix
Grünberger<sup>1</sup></a>, Michael Jüttner, Robert Knüppel,
<a href="https://orcid.org/0000-0002-0522-843X">Sébastien
Ferreira-Cerca<sup>2</sup></a>, and
<a href="https://orcid.org/0000-0002-0570-2517">Dina
Grohmann<sup>1</sup></a>  

<sup>1</sup> Department of Biochemistry, Genetics and Microbiology,
Institute of Microbiology, Single-Molecule Biochemistry Lab &
Biochemistry Centre Regensburg, University of Regensburg,
Universitätsstraße 31, 93053 Regensburg, Germany

<sup>2</sup> Biochemistry III – Institute for Biochemistry, Genetics and
Microbiology, University of Regensburg, Universitätsstraße 31, 93053
Regensburg, Germany.

<!-- README.md is generated from README.Rmd. Please edit that file -->

------------------------------------------------------------------------

## About this repository

This is the repository for the manuscript “Insights into rRNA processing
and modifications in Archaea using Nanopore-based RNA sequencing”.

The repository is currently actively developed.

[![Active
Development](https://img.shields.io/badge/Maintenance%20Level-Actively%20Developed-brightgreen.svg)](https://gist.github.com/cheerfulstoic/d107229326a01ff0f333a1d3476e068d)

## Full documentation here

<https://felixgrunberger.github.io/rRNA_maturation/>

## Preprint

This work is based on our previous preprint: [Exploring prokaryotic
transcription, operon structures, rRNA maturation and modifications
using Nanopore-based native RNA
sequencing.](%22https://www.biorxiv.org/content/10.1101/2019.12.18.880849v2.full%22)

## What can you find here

A description of the workflow using publicly available tools used to
basecall, demultiplex, trim and map (*direct cDNA*) data and data
preparation for modified base detection (*using direct RNA*) can be
found in the [pipeline](pipeline) section.

Downstream analysis, including  
- quality control  
- detection of rRNA processing sites and classification of rRNA
intermediates  
- Circular RNA detection  
- Modified base detection  
are based on custom [Rscripts](Rscripts).

## Data availability

### Sequencing files in original FAST5 format

Sequencing files XXX

### Basecalled and demultiplexed FASTQ files

For easier access, basecalled & demultiplexed FASTQ files are available

### Mapped BAM files

Minimap2-mapped untrimmed reads are also available

------------------------------------------------------------------------

## License

This project is under the general MIT License - see the
[LICENSE](LICENSE) file for details

## References
