<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Identification and trimming of full-length reads - rRNA maturation</title><link rel="shortcut icon" href="/rRNA_maturation/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/rRNA_maturation/assets/css/just-the-docs-default.css"> <script src="/rRNA_maturation/assets/js/vendor/lunr.min.js"></script> <script src="/rRNA_maturation/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Identification and trimming of full-length reads | rRNA maturation</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Identification and trimming of full-length reads" /><meta property="og:locale" content="en_US" /><meta name="description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><meta property="og:description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><link rel="canonical" href="http://localhost:4000/rRNA_maturation/processing_analysis/identification_full_length/" /><meta property="og:url" content="http://localhost:4000/rRNA_maturation/processing_analysis/identification_full_length/" /><meta property="og:site_name" content="rRNA maturation" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Identification and trimming of full-length reads" /> <script type="application/ld+json"> {"@type":"WebPage","url":"http://localhost:4000/rRNA_maturation/processing_analysis/identification_full_length/","headline":"Identification and trimming of full-length reads","description":"Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/rRNA_maturation/" class="site-title lh-tight"> rRNA maturation </a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="/rRNA_maturation/" class="nav-list-link">About</a><li class="nav-list-item"><a href="/rRNA_maturation/project" class="nav-list-link">Project</a><li class="nav-list-item"><a href="/rRNA_maturation/sample_preparation/" class="nav-list-link">Sample preparation</a><li class="nav-list-item"><a href="/rRNA_maturation/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Processing analysis category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/processing_analysis/" class="nav-list-link">Processing analysis</a><ul class="nav-list "><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/Basecalling/" class="nav-list-link">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/read_alignment/" class="nav-list-link">Read alignment</a></ul><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Modified base detection category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/mod_base/" class="nav-list-link">Modified base detection</a><ul class="nav-list "></ul><li class="nav-list-item"><a href="/rRNA_maturation/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="/rRNA_maturation/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search rRNA maturation" aria-label="Search rRNA maturation" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/rRNA_maturation" class="site-button" > Workflow on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><nav aria-label="Breadcrumb" class="breadcrumb-nav"><ol class="breadcrumb-nav-list"><li class="breadcrumb-nav-list-item"><a href="">Data analysis</a><li class="breadcrumb-nav-list-item"><span>Identification and trimming of full-length reads</span></ol></nav><div id="main-content" class="main-content" role="main"><h2 class="no_toc" id="trimming-of-reads-using-pychopper-cutadapt--samclip"> <a href="#trimming-of-reads-using-pychopper-cutadapt--samclip" class="anchor-heading" aria-labelledby="trimming-of-reads-using-pychopper-cutadapt--samclip"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Trimming of reads using <code class="language-plaintext highlighter-rouge">pychopper</code>, <code class="language-plaintext highlighter-rouge">cutadapt</code> &amp; <code class="language-plaintext highlighter-rouge">samclip</code></h2><hr /> <details open=""> <summary class="text-delta"> Table of contents </summary><ol id="markdown-toc"><li><a href="#identification-of-full-length-reads-using-pychopper" id="markdown-toc-identification-of-full-length-reads-using-pychopper">Identification of full-length reads using <code class="language-plaintext highlighter-rouge">pychopper</code></a><li><a href="#remove-polya-tails-using-cutadapt" id="markdown-toc-remove-polya-tails-using-cutadapt">Remove polyA-tails using <code class="language-plaintext highlighter-rouge">cutadapt</code></a><li><a href="#remove-remaining-ssp-adapter-using-cutadapt" id="markdown-toc-remove-remaining-ssp-adapter-using-cutadapt">Remove remaining SSP adapter using <code class="language-plaintext highlighter-rouge">cutadapt</code></a><li><a href="#mapping-of-trimmed-reads-removing-clips-using-samclip" id="markdown-toc-mapping-of-trimmed-reads-removing-clips-using-samclip">Mapping of trimmed reads, removing clips using <code class="language-plaintext highlighter-rouge">samclip</code></a></ol></details><hr /><h3 id="identification-of-full-length-reads-using-pychopper"> <a href="#identification-of-full-length-reads-using-pychopper" class="anchor-heading" aria-labelledby="identification-of-full-length-reads-using-pychopper"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Identification of full-length reads using <a href="https://github.com/nanoporetech/pychopper"><code class="language-plaintext highlighter-rouge">pychopper</code></a></h3><ul><li>Full-length cDNA reads containing SSP and VNP primers in the correct orientation were identified using <code class="language-plaintext highlighter-rouge">pychopper</code> (v.2.5.0) with standard parameters using the default pHMM backend and autotuned cutoff parameters estimated from subsampled data<li>Save output in <em>pychopper folder</em>.</ul><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/FASTQ/normal <span class="c"># input directory with all merged FASTQ files, 1 for each barcode or single DRS run</span>

<span class="c"># perform pychopper for all cDNA and (PCR)-cDNA files</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>.fastq
<span class="k">do</span> 

  <span class="c"># folder and filenames</span>
  <span class="nv">f_ex</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$f_ex</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">f_ex</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="c"># make directories</span>
  <span class="nb">mkdir </span>microbepore/data/pychopper/normal
  <span class="nb">mkdir </span>microbepore/data/pychopper/normal/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/pychopper/normal/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  <span class="c"># perform pychopper using precomputed q</span>
  cdna_classifier.py <span class="se">\</span>
  <span class="nt">-r</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_report.pdf <span class="se">\</span>
  <span class="nt">-t</span> 8 <span class="se">\</span>
  <span class="nt">-u</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified.fastq <span class="se">\</span>
  <span class="nt">-w</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_rescued.fastq <span class="se">\</span>
  <span class="nt">-S</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_stats.txt <span class="se">\</span>
  <span class="nv">$file</span> <span class="se">\</span>
  <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_full_length_output.fastq
<span class="k">done</span>
</code></pre></div></div><p>After a first round, a second round of <code class="language-plaintext highlighter-rouge">pychopper</code> was applied to the unclassified direct cDNA reads with DCS-specific read rescue enabled.</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/pychopper/normal <span class="c"># input directory with all merged FASTQ files, 1 for each barcode or single DRS run</span>

<span class="c"># perform pychopper using the -x rescue option for DCS files</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>unclassified.fastq <span class="c"># only use unclassified reads from first round as input</span>
<span class="k">do</span> 

  <span class="c"># folder and filenames</span>
  <span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="c"># make directories</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">dir</span><span class="k">}</span>/data/pychopper/rescued
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">dir</span><span class="k">}</span>/data/pychopper/rescued/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/pychopper/rescued/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>
  
  <span class="c"># perfrom pychopper using -X option for native cDNA datasets</span>
  cdna_classifier.py <span class="se">\</span>
  <span class="nt">-r</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_report.pdf <span class="se">\</span>
  <span class="nt">-t</span> 8 <span class="se">\</span>
  <span class="nt">-x</span> rescue <span class="se">\</span>
  <span class="nt">-u</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified.fastq <span class="se">\</span>
  <span class="nt">-w</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_rescued.fastq <span class="se">\</span>
  <span class="nt">-S</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_stats.txt <span class="se">\</span>
  <span class="nv">$file</span> <span class="se">\</span>
  <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_full_length_output.fastq
<span class="k">done</span>
</code></pre></div></div><p>Reads from rescued and normal folders were merged and used for subsequent steps.</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/pychopper/

<span class="c"># merge all full-length and rescued reads as full-length</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/normal/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>full_length_output.fastq <span class="c"># both normal and rescued folders</span>
<span class="k">do 
  </span><span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3,4,5<span class="si">)</span>

  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$foldername</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 2<span class="si">)</span> <span class="c"># get libary kit ID</span>
  
  <span class="nb">mkdir </span>microbepore/data/FASTQ/full_length
  <span class="nb">mkdir </span>microbepore/data/FASTQ/full_length/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/FASTQ/full_length/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>
  
  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$keyword</span> <span class="o">=</span>~ <span class="s2">"PCB109"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span><span class="nb">cat</span> <span class="nv">$file</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/normal/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_rescued.fastq <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_full_length_all.fastq
  <span class="k">elif</span> <span class="o">[[</span> <span class="nv">$keyword</span> <span class="o">=</span>~ <span class="s2">"DCS109"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span><span class="nb">cat</span> <span class="nv">$file</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/normal/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_rescued.fastq
    <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/rescued/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified_full_length_output.fastq
    <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/rescued/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_unclassified_rescued.fastq <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>_full_length_all.fastq
  <span class="k">fi
done</span>
</code></pre></div></div><p>For easier handling in the subsequent steps, DRS FASTQ files are also moved to the microbepore/data/FASTQ/full_length folder and adding <em>_full_length_all</em> to the filename.</p><h3 id="remove-polya-tails-using-cutadapt"> <a href="#remove-polya-tails-using-cutadapt" class="anchor-heading" aria-labelledby="remove-polya-tails-using-cutadapt"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Remove polyA-tails using <a href="https://cutadapt.readthedocs.io/en/stable/"><code class="language-plaintext highlighter-rouge">cutadapt</code></a></h3><ul><li>To evaluate the influence of different trimming approaches on the accuracy of transcript boundary analysis, we applied additional 5´ and 3´ trimming steps using <code class="language-plaintext highlighter-rouge">cutadapt</code> v3.2<li>To this end, polyA sequences were removed from the 3´ends:</ul><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/FASTQ/full_length <span class="c"># input directory with all merged FASTQ files, 1 for each barcode or single DRS run</span>

<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>_full_length_all.fastq
<span class="k">do</span> 

  <span class="c"># folder and filenames</span>
  <span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"."</span> <span class="nt">-f</span> 2<span class="si">)</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="nb">mkdir </span>microbepore/data/FASTQ/cutadapt
  <span class="nb">mkdir </span>microbepore/data/FASTQ/cutadapt/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/FASTQ/cutadapt/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>
  
  <span class="c"># cutadapt</span>
  cutadapt <span class="se">\</span>
    <span class="nt">-a</span> <span class="s2">"A{10}"</span> <span class="se">\ </span><span class="c"># trim polyAs longer than 10 bases from the 3´end</span>
    <span class="nt">-e</span> 1 <span class="se">\ </span><span class="c"># allowed error rate</span>
    <span class="nt">-j</span> 0 <span class="se">\ </span><span class="c"># auto-detect cores</span>
    <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.cutadapt.fastq <span class="se">\</span>
    <span class="k">${</span><span class="nv">file</span><span class="k">}</span>
<span class="k">done</span>
</code></pre></div></div><h3 id="remove-remaining-ssp-adapter-using-cutadapt"> <a href="#remove-remaining-ssp-adapter-using-cutadapt" class="anchor-heading" aria-labelledby="remove-remaining-ssp-adapter-using-cutadapt"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Remove remaining SSP adapter using <a href="https://cutadapt.readthedocs.io/en/stable/"><code class="language-plaintext highlighter-rouge">cutadapt</code></a></h3><ul><li>Remove remaining SSP sequences from the 5´ends of the cDNA reads using:</ul><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/FASTQ/cutadapt

<span class="c"># &gt;  SSP adapter</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>cutadapt.fastq
<span class="k">do 
  </span><span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"."</span> <span class="nt">-f</span> 2<span class="si">)</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="nb">mkdir </span>microbepore/data/FASTQ/cutadapt_SSP
  <span class="nb">mkdir </span>microbepore/data/FASTQ/cutadapt_SSP/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/FASTQ/cutadapt_SSP/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  cutadapt <span class="se">\</span>
    <span class="nt">-g</span> <span class="s2">"TTTCTGTTGGTGCTGATATTGCTGGG"</span> <span class="se">\</span>
    <span class="nt">-e</span> 1 <span class="se">\</span>
    <span class="nt">-j</span> 0 <span class="se">\</span>
    <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.cutadapt_SSP.fastq <span class="se">\</span>
    <span class="k">${</span><span class="nv">file</span><span class="k">}</span>
<span class="k">done</span>
</code></pre></div></div><h3 id="mapping-of-trimmed-reads-removing-clips-using-samclip"> <a href="#mapping-of-trimmed-reads-removing-clips-using-samclip" class="anchor-heading" aria-labelledby="mapping-of-trimmed-reads-removing-clips-using-samclip"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Mapping of trimmed reads, removing clips using <code class="language-plaintext highlighter-rouge">samclip</code></h3><ul><li>Finally, trimmed reads were mapped using <code class="language-plaintext highlighter-rouge">minimap2</code> as described before<li>Reads with more than 10 clipped bases on either side were removed from the alignments using <a href="https://github.com/tseemann/samclip"><code class="language-plaintext highlighter-rouge">samclip</code></a> (v.0.4.0)</ul><ol><li>Step: Align</ol><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/FASTQ/cutadapt_SSP
<span class="nv">fasta</span><span class="o">=</span>microbepore/data/genome/NC_000913.3.fasta <span class="c"># downloaded from GenBank</span>

<span class="c"># map (pychopper) &gt; polyA_trimmed &gt; SSP trimmed fastqs</span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>fastq
<span class="k">do 
  </span><span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="k">${</span><span class="nv">filename_extended</span><span class="k">}</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>

  <span class="nb">mkdir </span>microbepore/data/mapped/adapter_trimmed
  <span class="nb">mkdir </span>microbepore/data/mapped/adapter_trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/mapped/adapter_trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>

  <span class="c">## align using minimap2</span>
  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$filename</span> <span class="o">=</span>~ <span class="s2">"RNA"</span> <span class="o">]]</span><span class="p">;</span> 
  <span class="k">then</span>
  <span class="c"># align using minimap2</span>
  minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-uf</span> <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">fasta</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sam
  <span class="k">else
    </span>minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">fasta</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.sam
  <span class="k">fi
done</span>
</code></pre></div></div><ol><li>Step: Remove clipping &gt; 10 bases</ol><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/mapped/adapter_trimmed
<span class="nv">fasta</span><span class="o">=</span>microbepore/data/genome/NC_000913.3.fasta <span class="c"># downloaded from GenBank</span>
<span class="nv">transcripts</span><span class="o">=</span>microbepore/data/genomeNC_000913.3.transcripts.fasta <span class="c"># transcripts file made using gffread</span>

<span class="c"># remove reads with more than 10 bases that are clipped on either side. </span>
<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>.sam
<span class="k">do 
  </span><span class="nv">filename_extended</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">keyword</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"."</span> <span class="nt">-f</span> 2<span class="si">)</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="nv">$filename_extended</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">filename_extended</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$keyword</span> <span class="o">=</span>~ <span class="s2">"sam"</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then
    </span><span class="nb">echo</span> <span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
    <span class="nb">echo</span> <span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
    <span class="nb">echo</span> <span class="k">${</span><span class="nv">keyword</span><span class="k">}</span>
    
    <span class="nb">mkdir </span>microbepore/data/mapped/trimmed
    <span class="nb">mkdir </span>microbepore/data/mapped/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
    <span class="nv">output</span><span class="o">=</span>microbepore/data/mapped/trimmed/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
    <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>
  
    <span class="c"># remove mapped reads with a Maximum clip length to allow (10, 5 is default)</span>
    samclip <span class="nt">--max</span> 10 <span class="nt">--ref</span> <span class="k">${</span><span class="nv">fasta</span><span class="k">}</span> &lt; <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sam
    
    <span class="c"># convert to sorted.bam file</span>
    samtools flagstat <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sam <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.stats.txt
    samtools view <span class="nt">-bS</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.bam
    samtools <span class="nb">sort</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.bam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sorted.bam
    samtools index <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sorted.bam
    
    <span class="c">## remap fastq converted reads</span>
  bedtools bamtofastq <span class="nt">-i</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.clipped.sorted.bam <span class="nt">-fq</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.fastq
  
  <span class="c">## map again</span>
  <span class="k">if</span> <span class="o">[[</span> <span class="nv">$filename</span> <span class="o">=</span>~ <span class="s2">"RNA"</span> <span class="o">]]</span><span class="p">;</span> 
  <span class="k">then</span>
  <span class="c"># align using minimap2</span>
  minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-uf</span> <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">transcripts</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam
  <span class="k">else
    </span>minimap2 <span class="nt">-ax</span> splice <span class="nt">-p</span> 0.99 <span class="nt">-k14</span> <span class="nt">--MD</span> <span class="nt">-t</span> 8 <span class="k">${</span><span class="nv">transcripts</span><span class="k">}</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="o">&gt;</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam
  <span class="k">fi</span>
  
  <span class="c"># convert to sorted.bam file</span>
  samtools view <span class="nt">-bS</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.bam
  samtools <span class="nb">sort</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.bam <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sorted.bam
  samtools index <span class="k">${</span><span class="nv">output</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>.remapped.sorted.bam
  <span class="k">fi
done</span>
</code></pre></div></div><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2022 Felix Grünberger. Distributed by an <a href="https://github.com/felixgrunberger/rRNA_maturation/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
