<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Sample preparation - rRNA maturation</title><link rel="shortcut icon" href="/rRNA_maturation/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/rRNA_maturation/assets/css/just-the-docs-default.css"> <script src="/rRNA_maturation/assets/js/vendor/lunr.min.js"></script> <script src="/rRNA_maturation/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Sample preparation | rRNA maturation</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Sample preparation" /><meta property="og:locale" content="en_US" /><meta name="description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><meta property="og:description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><link rel="canonical" href="http://localhost:4000/rRNA_maturation/sample_preparation/" /><meta property="og:url" content="http://localhost:4000/rRNA_maturation/sample_preparation/" /><meta property="og:site_name" content="rRNA maturation" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Sample preparation" /> <script type="application/ld+json"> {"@type":"WebPage","url":"http://localhost:4000/rRNA_maturation/sample_preparation/","headline":"Sample preparation","description":"Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/rRNA_maturation/" class="site-title lh-tight"> rRNA maturation </a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="/rRNA_maturation/" class="nav-list-link">About</a><li class="nav-list-item"><a href="/rRNA_maturation/project" class="nav-list-link">Project</a><li class="nav-list-item active"><a href="/rRNA_maturation/sample_preparation/" class="nav-list-link active">Sample preparation</a><li class="nav-list-item"><a href="/rRNA_maturation/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Processing analysis category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/processing_analysis/" class="nav-list-link">Processing analysis</a><ul class="nav-list "><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/Basecalling/" class="nav-list-link">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/read_alignment/" class="nav-list-link">Read alignment</a></ul><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Modified base detection category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/mod_base/" class="nav-list-link">Modified base detection</a><ul class="nav-list "></ul><li class="nav-list-item"><a href="/rRNA_maturation/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="/rRNA_maturation/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search rRNA maturation" aria-label="Search rRNA maturation" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/rRNA_maturation" class="site-button" > Workflow on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><div id="main-content" class="main-content" role="main"><h2 class="no_toc" id="sample-preparation"> <a href="#sample-preparation" class="anchor-heading" aria-labelledby="sample-preparation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Sample preparation</h2><hr /> <details open=""> <summary class="text-delta"> Table of contents </summary><ol id="markdown-toc"><li><a href="#cell-growth-and-rna-extraction" id="markdown-toc-cell-growth-and-rna-extraction">Cell growth and RNA extraction</a><li><a href="#polya-tailing-rrna-depletion-and-additional-rna-treatment" id="markdown-toc-polya-tailing-rrna-depletion-and-additional-rna-treatment">Poly(A) tailing, rRNA depletion and additional RNA treatment</a><li><a href="#library-preparation" id="markdown-toc-library-preparation">Library preparation</a></ol></details><hr /><h3 id="cell-growth-and-rna-extraction"> <a href="#cell-growth-and-rna-extraction" class="anchor-heading" aria-labelledby="cell-growth-and-rna-extraction"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Cell growth and RNA extraction</h3><ul><li>Grow <em>Escherichia coli</em> K-12 MG1655 cells in rich medium (10 g tryptone, 5 g yeast extract, 5 g NaCl per liter, pH 7.2) to an OD<sub>600nm</sub> of 0.5-0.6 (or your organism of choice in the conditions you´re interested in)<li>Stabilize RNAs by adding two volumes of RNAlater (Thermo Fisher Scientific)<li>Store cultures at -20°C or harvest cells by centrifugation at 4°C<li>Extract RNA with the method of your choice. We (and others) had good experience with the RNeasy Kit from Qiagen<blockquote><p><strong>Note:</strong> Silica-membrane columns have a cut-off size of about 200 nucleotides</p></blockquote><li>Check integrity of total RNA (Bioanalyzer)<blockquote><p><strong>Note:</strong> RNA degradation has a massive influence on the results of your experiment</p></blockquote></ul><h3 id="polya-tailing-rrna-depletion-and-additional-rna-treatment"> <a href="#polya-tailing-rrna-depletion-and-additional-rna-treatment" class="anchor-heading" aria-labelledby="polya-tailing-rrna-depletion-and-additional-rna-treatment"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Poly(A) tailing, rRNA depletion and additional RNA treatment</h3><ul><li>Perform <strong>poly(A)–tailing</strong> using the <em>E. coli</em> poly(A) polymerase (New England Biolabs):<ul><li>Incubate RNAs at 70°C for 2 min and snap cool it on a pre-chilled freezer block<li>Incubate 5 µg RNA, 20 units poly(A) polymerase, 5 µl reaction buffer, 1 mM ATP for 15 min at 37°C in a total reaction volume of 50 µl.<li>Stop and clean up the reaction following the RNeasy Micro clean-up protocol (Qiagen)<li>Evaluate efficiency of poly(A)-tailing (Bioanalyzer, peaks of rRNAs)</ul></ul><blockquote><p><strong>Note:</strong> Efficiency of poly(A)-tailing is a critical step, but may not work equally well in other organisms.</p></blockquote><ul><li>Perform <strong>rRNA depletion</strong>, e.g. using the <a href="https://www.sitoolsbiotech.com/ribopools.php">Pan-Prokaryote riboPOOL by siTOOLs</a> (Clean-up can also be performed using the RNeasy protocol)</ul><blockquote><p><strong>Note:</strong> After depletion, only <strong>~5%</strong> of the initial RNA input quantity are left.</p></blockquote><ul><li>Perform additional treatment of your choice and clean-up, e.g. using the Terminator 5´-Phosphate-Dependent Exonuclease (TEX, Lucigen).</ul><blockquote><p><strong>Important</strong>: Before library preparation:</p><ol><li>Check extent of remaining buffer and DNA contamination that could have a negative impact on your library<li>Check RNA size and quantity (Qubit) to determine molarities as accurately as possible</ol></blockquote><h3 id="library-preparation"> <a href="#library-preparation" class="anchor-heading" aria-labelledby="library-preparation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Library preparation</h3><ul><li>Libraries for Nanopore sequencing were prepared from poly(A)-tailed RNAs according to the protocols provided by <a href="https://nanoporetech.com">ONT</a><li><strong>Critical steps</strong>:<ul><li>Input quality &amp; amount of RNA: Especially important for <strong>DRS</strong>, since you cannot get rid of excess amounts of RNA used as input. In that case, many transcripts that do not carry a motor protein are loaded onto the flowcell and overall have a negative impact on throughput/quality.<li>Reverse transcription: Maxima RT H Minus performs well and can be adjusted to high GC content<li>PCR for <strong>PCR-cDNA</strong> protocols: number of cycles adjusted to input amount (use minimal number cycles possible), extension time (adjust for longest transcripts you expect)<li>Beads to cDNA ratio (<strong>PCR-cDNA</strong>): Enrichment for longer transcripts using a 0.5 ratio<li>Molarity of ready-to-load library (check size &amp; quantity): Avoid overloading of your flowcells! We observed that especially small fragments lead to a poor translocation spped and quality of reads.</ul></ul><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2022 Felix Grünberger. Distributed by an <a href="https://github.com/felixgrunberger/rRNA_maturation/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
