<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Gene abundance estimation - microbepore</title><link rel="shortcut icon" href="/microbepore/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/microbepore/assets/css/just-the-docs-default.css"> <script type="text/javascript" src="/microbepore/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="/microbepore/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Gene abundance estimation | microbepore</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Gene abundance estimation" /><meta property="og:locale" content="en_US" /><meta name="description" content="Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules" /><meta property="og:description" content="Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules" /><link rel="canonical" href="http://localhost:4000/microbepore/data_analysis/quantification/" /><meta property="og:url" content="http://localhost:4000/microbepore/data_analysis/quantification/" /><meta property="og:site_name" content="microbepore" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Gene abundance estimation" /> <script type="application/ld+json"> {"description":"Transcriptome analysis of prokaryotes based on Nanopore sequencing of RNA and cDNA molecules","headline":"Gene abundance estimation","publisher":{"@type":"Organization","logo":{"@type":"ImageObject","url":"http://localhost:4000/microbepore/assets/images/microbepore_logo.png"}},"url":"http://localhost:4000/microbepore/data_analysis/quantification/","@type":"WebPage","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/microbepore/" class="site-title lh-tight"><div class="site-logo"></div></a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="http://localhost:4000/microbepore/" class="nav-list-link">About</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/project" class="nav-list-link">Project</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/sample_preparation/" class="nav-list-link">Sample preparation</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item active"><a href="#" class="nav-list-expander"><svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg></a><a href="http://localhost:4000/microbepore/data_analysis/" class="nav-list-link">Data analysis</a><ul class="nav-list "><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/project_organization/" class="nav-list-link">Project organisation</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/Basecalling/" class="nav-list-link">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/read_alignment/" class="nav-list-link">Read alignment</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/identification_full_length/" class="nav-list-link">Identification and trimming of full-length reads</a><li class="nav-list-item active"><a href="http://localhost:4000/microbepore/data_analysis/quantification/" class="nav-list-link active">Gene abundance estimation</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/qc_raw_reads/" class="nav-list-link">Quality control</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/transcript_boundaries/" class="nav-list-link">Detection of transcript boundaries</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/gene_body/" class="nav-list-link">Gene body coverage</a><li class="nav-list-item "><a href="http://localhost:4000/microbepore/data_analysis/operon_detection/" class="nav-list-link">Transcriptional unit analysis</a></ul><li class="nav-list-item"><a href="http://localhost:4000/microbepore/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="http://localhost:4000/microbepore/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search microbepore" aria-label="Search microbepore" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/microbepore" class="site-button" > microbepore on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><nav aria-label="Breadcrumb" class="breadcrumb-nav"><ol class="breadcrumb-nav-list"><li class="breadcrumb-nav-list-item"><a href="http://localhost:4000/microbepore/data_analysis/">Data analysis</a><li class="breadcrumb-nav-list-item"><span>Gene abundance estimation</span></ol></nav><div id="main-content" class="main-content" role="main"><h2 id="gene-abundance-estimation"> <a href="#gene-abundance-estimation" class="anchor-heading" aria-labelledby="gene-abundance-estimation"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Gene abundance estimation</h2><ul><li>To estimate gene abundances <code class="language-plaintext highlighter-rouge">salmon</code> (v.1.4.0) was applied in alignment-based mode as described in https://salmon.readthedocs.io/en/latest/salmon.html#quantifying-in-alignment-based-mode<li>For installation you can follow the explanations listed at https://combine-lab.github.io/salmon/getting_started/ (installs salmon in own conda environment)</ul><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">input</span><span class="o">=</span>microbepore/data/mapped/raw <span class="c"># input directory with all remapped files</span>

<span class="k">for </span>file <span class="k">in</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span>/<span class="k">*</span>/<span class="k">*</span>/<span class="k">*</span>remapped.sorted.bam
<span class="k">do</span>
  
  <span class="c"># folder and filenames</span>
  <span class="nv">f_ex</span><span class="o">=</span><span class="k">${</span><span class="nv">file</span><span class="p">##*/</span><span class="k">}</span>
  <span class="nv">foldername</span><span class="o">=</span><span class="si">$(</span><span class="nb">echo</span> <span class="k">${</span><span class="nv">f_ex</span><span class="k">}</span> | <span class="nb">cut</span> <span class="nt">-d</span><span class="s2">"_"</span> <span class="nt">-f</span> 1,2,3<span class="si">)</span>
  <span class="nv">filename</span><span class="o">=</span><span class="k">${</span><span class="nv">f_ex</span><span class="p">%%.*</span><span class="k">}</span>
  
  <span class="c"># create dir for quantification using salmon in alignment-based mode (e.g. used in conda environment)</span>
  <span class="nb">mkdir </span>microbepore/data/salmon
  <span class="nb">mkdir </span>microbepore/data/salmon/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>
  <span class="nv">output</span><span class="o">=</span>microbepore/data/salmon/<span class="k">${</span><span class="nv">foldername</span><span class="k">}</span>/<span class="k">${</span><span class="nv">filename</span><span class="k">}</span>
  <span class="nb">mkdir</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span>
  
  conda activate salmon <span class="c"># activate conda environment</span>
  
  <span class="c"># use salmon in alignment-based mode</span>
  salmon quant <span class="se">\</span>
  <span class="nt">-t</span> <span class="k">${</span><span class="nv">transcripts</span><span class="k">}</span> <span class="se">\</span>
  <span class="nt">-l</span> A <span class="se">\</span>
  <span class="nt">-a</span> <span class="k">${</span><span class="nv">file</span><span class="k">}</span> <span class="se">\</span>
  <span class="nt">-o</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span> <span class="se">\</span>
  <span class="nt">--threads</span> 8 
  
  conda deactivate
  
<span class="k">done</span>
</code></pre></div></div><ul><li>Transcripts per million (TPM) were re-calculated using the salmon-computed effective transcript length, after dropping reads mapping to rRNAs, that are variable between non-depleted and depleted RNA sets (compare custom <a href="https://github.com/felixgrunberger/microbepore/blob/master/Rscripts/salmon_analysis.R">Rscripts/salmon_analysis.R</a>).<li>Calculation of custom TPMs is performed in the <code class="language-plaintext highlighter-rouge">modify_salmon_output</code> function:</ul><div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">modify_salmon_output</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">input</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="p">){</span><span class="w">
  </span><span class="c1"># read in salmon output</span><span class="w">
  </span><span class="n">suppressMessages</span><span class="p">(</span><span class="n">vroom</span><span class="p">(</span><span class="n">input</span><span class="p">,</span><span class="w"> </span><span class="n">num_threads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">  
    </span><span class="c1"># sort by Number of reads mapping</span><span class="w">
    </span><span class="n">arrange</span><span class="p">(</span><span class="n">desc</span><span class="p">(</span><span class="n">NumReads</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># read in salmon output</span><span class="w">
    </span><span class="c1"># rename columns</span><span class="w">
    </span><span class="n">mutate</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">NumReads</span><span class="p">,</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Name</span><span class="p">,</span><span class="w"> </span><span class="n">salmon_tpm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TPM</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># drop all rRNAs</span><span class="w">
    </span><span class="n">dplyr</span><span class="o">::</span><span class="n">filter</span><span class="p">(</span><span class="o">!</span><span class="n">gene</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">names_rRNA</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># select columns</span><span class="w">
    </span><span class="n">select</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">salmon_tpm</span><span class="p">,</span><span class="n">EffectiveLength</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># rename genes</span><span class="w">
    </span><span class="n">mutate</span><span class="p">(</span><span class="n">gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">str_split_fixed</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span><span class="s2">"-"</span><span class="p">,</span><span class="m">2</span><span class="p">)[,</span><span class="m">2</span><span class="p">])</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># sort by number of counts</span><span class="w">
    </span><span class="n">arrange</span><span class="p">(</span><span class="n">desc</span><span class="p">(</span><span class="n">counts</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># add gene information (including with to table)</span><span class="w">
    </span><span class="n">left_join</span><span class="p">(</span><span class="n">ecoli_gff</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"gene"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># make RNA type categories</span><span class="w">
    </span><span class="n">mutate</span><span class="p">(</span><span class="n">type_fine</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">type</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"rRNA"</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">locus_name</span><span class="p">),</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">type</span><span class="p">)),</span><span class="w">
           </span><span class="n">type_fine</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">type</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"tRNA"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ncRNA"</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">type_fine</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># calculate RPK from salmon-calculated effectiveLength, TPM(hand) = RPK/sum(RPK)/1000000, additionally calc RPKM       </span><span class="w">
    </span><span class="n">mutate</span><span class="p">(</span><span class="n">sample</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">method</span><span class="p">,</span><span class="w">
           </span><span class="n">rpk</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span><span class="n">counts</span><span class="o">/</span><span class="n">EffectiveLength</span><span class="o">*</span><span class="m">1000</span><span class="p">),</span><span class="w">
           </span><span class="n">TPM_hand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rpk</span><span class="o">/</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">rpk</span><span class="p">,</span><span class="w"> </span><span class="n">na.rm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="o">/</span><span class="m">1000000</span><span class="p">),</span><span class="w">
           </span><span class="n">rpkm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span><span class="n">counts</span><span class="o">/</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span><span class="o">/</span><span class="m">1000000</span><span class="p">))</span><span class="o">/</span><span class="n">width</span><span class="o">*</span><span class="m">1000</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="c1"># TPM_hand is the value used        </span><span class="w">
    </span><span class="n">dplyr</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">type_fine</span><span class="p">,</span><span class="w"> </span><span class="n">counts</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="p">,</span><span class="n">rpkm</span><span class="p">,</span><span class="w"> </span><span class="n">TPM_hand</span><span class="p">,</span><span class="w"> </span><span class="n">salmon_tpm</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2021 Felix Gr√ºnberger. Distributed by an <a href="https://github.com/felixgrunberger/microbepore/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
