<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title>Basecalling and demultiplexing of raw reads - rRNA maturation</title><link rel="shortcut icon" href="/rRNA_maturation/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="/rRNA_maturation/assets/css/just-the-docs-default.css"> <script src="/rRNA_maturation/assets/js/vendor/lunr.min.js"></script> <script src="/rRNA_maturation/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Basecalling and demultiplexing of raw reads | rRNA maturation</title><meta name="generator" content="Jekyll v3.9.0" /><meta property="og:title" content="Basecalling and demultiplexing of raw reads" /><meta property="og:locale" content="en_US" /><meta name="description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><meta property="og:description" content="Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing" /><link rel="canonical" href="http://localhost:4000/rRNA_maturation/processing_analysis/Basecalling/" /><meta property="og:url" content="http://localhost:4000/rRNA_maturation/processing_analysis/Basecalling/" /><meta property="og:site_name" content="rRNA maturation" /><meta name="twitter:card" content="summary" /><meta property="twitter:title" content="Basecalling and demultiplexing of raw reads" /> <script type="application/ld+json"> {"@type":"WebPage","url":"http://localhost:4000/rRNA_maturation/processing_analysis/Basecalling/","headline":"Basecalling and demultiplexing of raw reads","description":"Insights into rRNA processing and modifications in Archaea using Nanopore-based RNA sequencing","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="svg-link" viewBox="0 0 24 24"><title>Link</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-link"><path d="M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71"></path><path d="M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71"></path> </svg> </symbol> <symbol id="svg-search" viewBox="0 0 24 24"><title>Search</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-search"> <circle cx="11" cy="11" r="8"></circle><line x1="21" y1="21" x2="16.65" y2="16.65"></line> </svg> </symbol> <symbol id="svg-menu" viewBox="0 0 24 24"><title>Menu</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"></line><line x1="3" y1="6" x2="21" y2="6"></line><line x1="3" y1="18" x2="21" y2="18"></line> </svg> </symbol> <symbol id="svg-arrow-right" viewBox="0 0 24 24"><title>Expand</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-chevron-right"><polyline points="9 18 15 12 9 6"></polyline> </svg> </symbol> <symbol id="svg-doc" viewBox="0 0 24 24"><title>Document</title><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file"><path d="M13 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V9z"></path><polyline points="13 2 13 9 20 9"></polyline> </svg> </symbol> </svg><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000/rRNA_maturation/" class="site-title lh-tight"> rRNA maturation </a> <a href="#" id="menu-button" class="site-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-menu"></use></svg> </a></div><nav role="navigation" aria-label="Main" id="site-nav" class="site-nav"><ul class="nav-list"><li class="nav-list-item"><a href="/rRNA_maturation/" class="nav-list-link">About</a><li class="nav-list-item"><a href="/rRNA_maturation/project" class="nav-list-link">Project</a><li class="nav-list-item"><a href="/rRNA_maturation/sample_preparation/" class="nav-list-link">Sample preparation</a><li class="nav-list-item"><a href="/rRNA_maturation/sequencing/" class="nav-list-link">Sequencing</a><li class="nav-list-item active"><a href="#" class="nav-list-expander" aria-label="toggle links in Processing analysis category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/processing_analysis/" class="nav-list-link">Processing analysis</a><ul class="nav-list "><li class="nav-list-item active"><a href="/rRNA_maturation/processing_analysis/Basecalling/" class="nav-list-link active">Basecalling and demultiplexing of raw reads</a><li class="nav-list-item "><a href="/rRNA_maturation/processing_analysis/read_alignment/" class="nav-list-link">Read alignment</a></ul><li class="nav-list-item"><a href="#" class="nav-list-expander" aria-label="toggle links in Modified base detection category"> <svg viewBox="0 0 24 24"><use xlink:href="#svg-arrow-right"></use></svg> </a><a href="/rRNA_maturation/mod_base/" class="nav-list-link">Modified base detection</a><ul class="nav-list "></ul><li class="nav-list-item"><a href="/rRNA_maturation/data/" class="nav-list-link">Data</a><li class="nav-list-item"><a href="/rRNA_maturation/contact/" class="nav-list-link">Contact</a></ul></nav><footer class="site-footer"> This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</footer></div><div class="main" id="top"><div id="main-header" class="main-header"><div class="search"><div class="search-input-wrap"> <input type="text" id="search-input" class="search-input" tabindex="0" placeholder="Search rRNA maturation" aria-label="Search rRNA maturation" autocomplete="off"> <label for="search-input" class="search-label"><svg viewBox="0 0 24 24" class="search-icon"><use xlink:href="#svg-search"></use></svg></label></div><div id="search-results" class="search-results"></div></div><nav aria-label="Auxiliary" class="aux-nav"><ul class="aux-nav-list"><li class="aux-nav-list-item"> <a href="//github.com/felixgrunberger/rRNA_maturation" class="site-button" > Workflow on GitHub </a></ul></nav></div><div id="main-content-wrap" class="main-content-wrap"><nav aria-label="Breadcrumb" class="breadcrumb-nav"><ol class="breadcrumb-nav-list"><li class="breadcrumb-nav-list-item"><a href="http://localhost:4000/rRNA_maturation/processing_analysis/">Processing analysis</a><li class="breadcrumb-nav-list-item"><span>Basecalling and demultiplexing of raw reads</span></ol></nav><div id="main-content" class="main-content" role="main"><h2 class="no_toc" id="basecalling-and-demultiplexing-of-raw-reads"> <a href="#basecalling-and-demultiplexing-of-raw-reads" class="anchor-heading" aria-labelledby="basecalling-and-demultiplexing-of-raw-reads"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Basecalling and demultiplexing of raw reads</h2><hr /> <details open=""> <summary class="text-delta"> Table of contents </summary><ol id="markdown-toc"><li><a href="#basecalling-of-raw-reads-using-guppy_basecaller" id="markdown-toc-basecalling-of-raw-reads-using-guppy_basecaller">Basecalling of raw reads using <code class="language-plaintext highlighter-rouge">guppy_basecaller</code></a><li><a href="#demultiplexing-of-basecalled-reads-using-guppy_barcoder" id="markdown-toc-demultiplexing-of-basecalled-reads-using-guppy_barcoder">Demultiplexing of basecalled reads using <code class="language-plaintext highlighter-rouge">guppy_barcoder</code></a></ol></details><hr /><ul><li>For better comparability we re-basecalled the FAST5 file from all runs using the same version on guppy on a Mk1C (ont-guppy-for-mk1c v4.3.4)<li>Once the Mk1C is set up correctly and connected to the internet, updates are displayed automatically and can be completed by click<li>Other software downloads are available from the <a href="https://community.nanoporetech.com/downloads">ONT downloads page</a></ul><h3 id="basecalling-of-raw-reads-using-guppy_basecaller"> <a href="#basecalling-of-raw-reads-using-guppy_basecaller" class="anchor-heading" aria-labelledby="basecalling-of-raw-reads-using-guppy_basecaller"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Basecalling of raw reads using <code class="language-plaintext highlighter-rouge">guppy_basecaller</code></h3><ul><li>After sequencing (and despite live-basecalling) all datasets in the <em>raw_FAST5 folder</em> were re-basecalled using <code class="language-plaintext highlighter-rouge">guppy</code> (ont-guppy-for-mk1c v4.3.4) in high-accuracy mode (rna_r9.4.1_70bps_hac.cfg, dna_r9.4.1_450bps_hac.cfg) without quality filtering<li>The output files in FASTQ format were written to the <em>basecalled folder</em>.</ul><p><strong>Note</strong>: Mk1C</p><ul><li><a href="https://community.nanoporetech.com/posts/fastest-way-to-re-basecall">We (and others)</a> noticed the apparently acces via GUI is not the fastest way to perform basecalling<li>In case you want to optimise GPU-accelerated basecalling on the Mk1C (settings for NVIDIA Jetson TX2) using have a look at <a href="https://github.com/sirselim/jetson_nanopore_sequencing">Miles Benton`s GitHub</a></ul><p>According this his suggestions (which really speed up the process), we performed basecalling like this (using ssh access to the Mk1C):</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/raw_FAST5/run_id <span class="c"># add run id</span>
<span class="nv">output_DRS</span><span class="o">=</span>microbepore/data/FASTQ/normal/run_id <span class="c"># add run id</span>
<span class="nv">output_cDNA</span><span class="o">=</span>microbepore/data/basecalled/run_id <span class="c"># add run id</span>

<span class="c"># Basecalling of DRS files</span>
guppy_basecaller <span class="se">\</span>
<span class="nt">--input_path</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span> <span class="se">\ </span><span class="c"># input path</span>
<span class="nt">--save_path</span> <span class="k">${</span><span class="nv">output_DRS</span><span class="k">}</span> <span class="se">\ </span><span class="c"># output path</span>
<span class="nt">-c</span> rna_r9.4.1_70bps_hac.cfg  <span class="se">\ </span><span class="c"># config file: high accuracy RNA</span>
<span class="nt">--calib_detect</span> <span class="se">\ </span><span class="c"># detect calibration spike-in</span>
<span class="nt">--reverse_sequence</span> <span class="nb">true</span> <span class="se">\ </span><span class="c"># reverse since sequenced 3´--&gt;5´</span>
<span class="nt">--u_substitution</span> <span class="nb">true</span> <span class="se">\ </span><span class="c"># replace U´s with T´s</span>
<span class="nt">--compress_fastq</span> <span class="se">\ </span><span class="c"># compress output</span>
<span class="nt">--fast5_out</span> <span class="se">\ </span><span class="c"># output FAST5</span>
<span class="nt">--recursive</span> <span class="se">\ </span><span class="c"># look for FAST5 recursively in path</span>
<span class="nt">--progress_stats_frequency</span> 60 <span class="se">\ </span><span class="c"># output progress every minute</span>
<span class="nt">--chunks_per_runner</span> 256 <span class="se">\ </span><span class="c"># options for Mk1C</span>
<span class="nt">--gpu_runners_per_device</span> 4 <span class="se">\ </span><span class="c"># options for Mk1C</span>
<span class="nt">--num_callers</span> 1 <span class="se">\ </span><span class="c"># options for Mk1C</span>
<span class="nt">-x</span> auto <span class="c"># options for Mk1C</span>

<span class="c"># Basecalling of cDNA files </span>
guppy_basecaller <span class="se">\</span>
<span class="nt">--input_path</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span> <span class="se">\</span>
<span class="nt">--save_path</span> <span class="k">${</span><span class="nv">output_cDNA</span><span class="k">}</span> <span class="se">\</span>
<span class="nt">-c</span> dna_r9.4.1_450bps_hac.cfg <span class="se">\ </span><span class="c"># config file: high accuracy cDNA </span>
<span class="nt">--compress_fastq</span> <span class="se">\</span>
<span class="nt">--fast5_out</span> <span class="se">\</span>
<span class="nt">--recursive</span> <span class="se">\</span>
<span class="nt">--progress_stats_frequency</span> 60 <span class="se">\</span>
<span class="nt">--chunks_per_runner</span> 256 <span class="se">\</span>
<span class="nt">--gpu_runners_per_device</span> 4 <span class="se">\</span>
<span class="nt">--num_callers</span> 1 <span class="se">\</span>
<span class="nt">-x</span> auto
</code></pre></div></div><p><strong>Note</strong>:</p><ul><li>DRS &amp; (PCR-)cDNA runs require different options.<li>Config file selection based on selected accuracy, flowcell version, library preparation kit are listed with <code class="language-plaintext highlighter-rouge">guppy_basecaller --print_workflows</code></ul><p>Since basecalling is probably taking quite a while, but you want to keep the process running without interruption you can use the <a href="https://linuxhint.com/how_to_use_nohup_linux/"><code class="language-plaintext highlighter-rouge">nohup</code></a> command. You can use it like <code class="language-plaintext highlighter-rouge">nohup guppy_basecaller ..... &amp;</code>. This starts basecalling using <code class="language-plaintext highlighter-rouge">guppy</code> and puts to process in the background (that´s what the <code class="language-plaintext highlighter-rouge">&amp;</code> does). Now, you can log out from the Mk1C and the process keeps running. To monitor the process you can check the shell output that is written to <code class="language-plaintext highlighter-rouge">nohup.out</code>.</p><p>With the selected options <code class="language-plaintext highlighter-rouge">guppy</code> produces fast5_pass, fast5_fail, fastq, summary and report files that are written to the <em>FASTQ folder</em>. FASTQ are not grouped in pass and fail groups since <code class="language-plaintext highlighter-rouge">--min_qscore</code> is not enabled. Multiple FASTQs can be merged using <code class="language-plaintext highlighter-rouge">cat microbepore/data/basecalled/run_id/*.fastq &gt; microbepore/data/basecalled/run_id/run_id.fastq</code>.</p><p>Sequencing summary files are also written to the <em>FASTQ folder</em> and are used during the quality control of the runs and reads. For better viewing they can be moved to the <em>summary folder</em> using <code class="language-plaintext highlighter-rouge">mv microbepore/data/FASTQ/run_id/sequencing_summary.txt microbepore/data/summary/run_id.txt</code></p><h3 id="demultiplexing-of-basecalled-reads-using-guppy_barcoder"> <a href="#demultiplexing-of-basecalled-reads-using-guppy_barcoder" class="anchor-heading" aria-labelledby="demultiplexing-of-basecalled-reads-using-guppy_barcoder"><svg viewBox="0 0 16 16" aria-hidden="true"><use xlink:href="#svg-link"></use></svg></a> Demultiplexing of basecalled reads using <code class="language-plaintext highlighter-rouge">guppy_barcoder</code></h3><p>Next, multiplexed cDNA libraries are demultiplexed in a separate step using <code class="language-plaintext highlighter-rouge">guppy_barcoder</code>.</p><div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># files</span>
<span class="nv">input</span><span class="o">=</span>microbepore/data/basecalled/run_id <span class="c"># add run id</span>
<span class="nv">output</span><span class="o">=</span>microbepore/data/FASTQ/normal/run_id <span class="c"># add run id</span>

<span class="c"># Demultiplexing of (PCR-)cDNA files</span>
guppy_barcoder <span class="se">\</span>
<span class="nt">--input_path</span> <span class="k">${</span><span class="nv">input</span><span class="k">}</span> <span class="se">\</span>
<span class="nt">--save_path</span> <span class="k">${</span><span class="nv">output</span><span class="k">}</span> <span class="se">\</span>
<span class="nt">--config</span> configuration.cfg <span class="se">\</span>
<span class="nt">--barcode_kits</span> SQK-PCB109 <span class="se">\</span>
<span class="nt">--progress_stats_frequency</span> 60
</code></pre></div></div><p>Multiple FASTQs are written to the <em>FASTQ folder</em> and can be merged with e.g. <code class="language-plaintext highlighter-rouge">cat microbepore/data/FASTQ/run_id/barcode01/*.fastq &gt; microbepore/data/FASTQ/run_id/run_id_barcode01.fastq</code>. Barcode summary files are written to the <em>FASTQ folder</em> and can be moved to the <em>barcode folder</em>.</p><hr><footer><p><a href="#top" id="back-to-top">Back to top</a></p><p class="text-small text-grey-dk-100 mb-0">Copyright &copy; 2022 Felix Grünberger. Distributed by an <a href="https://github.com/felixgrunberger/rRNA_maturation/tree/master/LICENSE.txt">MIT license.</a></p><div class="d-flex mt-2"></div></footer></div></div><a href="#" id="search-button" class="search-button"> <svg viewBox="0 0 24 24" class="icon"><use xlink:href="#svg-search"></use></svg> </a><div class="search-overlay"></div></div>
